cpp/tests/sparse/preprocess.cu

```diff
@@ -0,0 +1,299 @@
+{
+  cudaStream_t stream  = raft::resource::get_cuda_stream(handle);
+  auto coo_struct_view = raft::make_device_coordinate_structure_view(
+    rows.data_handle(), columns.data_handle(), num_rows, num_cols, int(rows.size()));
+  auto c_matrix = raft::make_device_coo_matrix<T2, T1, T1, T1>(handle, coo_struct_view);
+  raft::update_device<T2>(
+    c_matrix.view().get_elements().data(), values.data_handle(), int(values.size()), stream);
+  return c_matrix;
+}
+
+template <typename T2, typename T1>
+raft::device_coo_matrix<T2, T1, T1, T1, raft::device_uvector_policy, raft::PRESERVING>
+create_coo_matrix(raft::resources& handle, raft::sparse::COO<T2, T1, T1>& coo)
+{
+  cudaStream_t stream  = raft::resource::get_cuda_stream(handle);
+  auto coo_struct_view = raft::make_device_coordinate_structure_view(
+    coo.rows(), coo.cols(), coo.n_rows, coo.n_cols, int(coo.nnz));
+  auto c_matrix = raft::make_device_coo_matrix<T2, T1, T1, T1>(handle, coo_struct_view);
+  raft::update_device<T2>(c_matrix.view().get_elements().data(), coo.vals(), coo.nnz, stream);
+  return c_matrix;
+}
+
+template <typename Type_f, typename Index_>
+struct SparsePreprocessInputs {
+  int n_rows;
+  int n_cols;
+  int nnz_edges;
+};
+
+template <typename Type_f, typename Index_>
+class SparsePreprocessCSR
+  : public ::testing::TestWithParam<SparsePreprocessInputs<Type_f, Index_>> {
+ public:
+  SparsePreprocessCSR()
+    : params(::testing::TestWithParam<SparsePreprocessInputs<Type_f, Index_>>::GetParam()),
+      stream(resource::get_cuda_stream(handle))
+  {
+  }
+
+ protected:
+  void SetUp() override {}
+
+  void Run(bool bm25_on, bool coo_on)
+  {
+    cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+    int num_rows        = pow(2, params.n_rows);
+    int num_cols        = pow(2, params.n_cols);
+    int nnz             = params.nnz_edges;
+    auto a_rows         = raft::make_device_vector<Index_, int64_t>(handle, nnz);
+    auto a_columns      = raft::make_device_vector<Index_, int64_t>(handle, nnz);
+    auto a_values       = raft::make_device_vector<Type_f, int64_t>(handle, nnz);
+
+    raft::util::create_dataset<Index_, Type_f>(
+      handle, a_rows.view(), a_columns.view(), a_values.view(), 5, params.n_rows, params.n_cols);
+
+    auto b_rows    = raft::make_device_vector<Index_, int64_t>(handle, nnz);
+    auto b_columns = raft::make_device_vector<Index_, int64_t>(handle, nnz);
+    auto b_values  = raft::make_device_vector<Type_f, int64_t>(handle, nnz);
+    raft::util::create_dataset<Index_, Type_f>(handle,
+                                               b_rows.view(),
+                                               b_columns.view(),
+                                               b_values.view(),
+                                               5,
+                                               params.n_rows,
+                                               params.n_cols,
+                                               67584);
```

Comment:
Fixed random number seeds will cause the generated test data to always be the same, which may not cover certain special input situations and thus miss potential bugs.